okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java

```diff
@@ -458,14 +487,57 @@ private static void addOrReplace(List<TenantModuleDescriptor> tml, ModuleDescrip
     tml.add(t);
   }
 
+  private static void upgradeLeafs(ModuleDescriptor md, Map<String, ModuleDescriptor> modsAvailable,
+    Map<String, ModuleDescriptor> modsEnabled, List<TenantModuleDescriptor> tml) {
+
+    Iterator<ModuleDescriptor> it = modsEnabled.values().iterator();
+    while (it.hasNext()) {
+      ModuleDescriptor me = it.next();
+      if (me.equals(md)) {
+        continue;
+      }
+      ModuleDescriptor mTo = null;
+      for (InterfaceDescriptor prov : md.getProvidesList()) {
+        for (InterfaceDescriptor req : me.getRequiresOptionalList()) {
+          if (prov.getId().equals(req.getId()) && !prov.isCompatible(req)) {
+            for (ModuleDescriptor ma : modsAvailable.values()) {
+              if (me.getProduct().equals(ma.getProduct())) {
+                for (InterfaceDescriptor re1 : ma.getRequiresOptionalList()) {
+                  if (prov.isCompatible(re1)) {
+                    if (mTo == null || ma.compareTo(mTo) > 0) {
+                      mTo = ma;
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
```

Comment:
This will reduce the cyclomatic complexity from 52 to 37 -- less nesting improves readability:

      for (InterfaceDescriptor prov : md.getProvidesList()) {
        for (InterfaceDescriptor req : me.getRequiresOptionalList()) {
          if (! prov.getId().equals(req.getId()) || prov.isCompatible(req)) {
            continue;
          }
          for (ModuleDescriptor ma : modsAvailable.values()) {
            if (! me.getProduct().equals(ma.getProduct())) {
              continue;
            }
            for (InterfaceDescriptor re1 : ma.getRequiresOptionalList()) {
              if (prov.isCompatible(re1) && (mTo == null || ma.compareTo(mTo) > 0)) {
                mTo = ma;
              }
            }
          }
        }
      }

Reply:
Done